#Homework 4 - Mahima Regunathan#

Colloboration: Alex Rose

*** Task 1. Imports ***

In [4]:
import pyspark.sql.functions as F

*** Task 2. Explore the data folder***

In [6]:
dbutils.fs.ls("dbfs:/databricks-datasets/definitive-guide/data/activity-data/")

Out[ 2 ]: [FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/_SUCCESS', name='_SUCCESS', size=0),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/_committed_730451297822678341', name='_committed_730451297822678341', size=6824),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/_started_730451297822678341', name='_started_730451297822678341', size=0),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00000-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00000-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831354),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00001-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00001-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831559),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00002-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00002-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831751),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00003-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00003-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831458),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00004-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00004-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831646),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00005-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00005-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831601),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00006-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00006-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831857),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00007-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00007-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831481),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00008-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00008-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831253),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00009-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00009-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831585),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00010-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00010-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831941),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00011-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00011-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831650),
 FileInfo(path='dbfs:/databricks-datasets/definitive-guide/data/activity-data/part-00012-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', name='part-00012-tid-730451297822678341-1dda7027-2071-4d73-a0e2-7fb6a91e1d1f-0-c000.json', size=14831577),
 FileInfo(path='dbfs:/databricks-datasets/definitive

There are 80 json files in the above path (Found it using .json in the file name and it ran from part-00000 to 00079)

***Task 3. Read the data***

In [9]:
staticDF = spark.read.json("dbfs:/databricks-datasets/definitive-guide/data/activity-data/")

In [10]:
dataSchema = staticDF.schema
dataSchema

Out[ 4 ]: StructType(List(StructField(Arrival_Time,LongType,true),StructField(Creation_Time,LongType,true),StructField(Device,StringType,true),StructField(Index,LongType,true),StructField(Model,StringType,true),StructField(User,StringType,true),StructField(gt,StringType,true),StructField(x,DoubleType,true),StructField(y,DoubleType,true),StructField(z,DoubleType,true)))

In [11]:
staticDF.printSchema()

root
-- Arrival_Time: long (nullable = true)
-- Creation_Time: long (nullable = true)
-- Device: string (nullable = true)
-- Index: long (nullable = true)
-- Model: string (nullable = true)
-- User: string (nullable = true)
-- gt: string (nullable = true)
-- x: double (nullable = true)
-- y: double (nullable = true)
-- z: double (nullable = true)

***Task 4. View the data and obtain some descriptive stats***

***a***

In [14]:
display(staticDF.head(15))

Arrival_Time Creation_Time Device Index Model User gt x y z 1424686735090 1424686733090638193 nexus4_1 18 nexus4 g stand 3.356934E-4 -5.645752E-4 -0.018814087 1424686735292 1424688581345918092 nexus4_2 66 nexus4 g stand -0.005722046 0.029083252 0.005569458 1424686735500 1424686733498505625 nexus4_1 99 nexus4 g stand 0.0078125 -0.017654419 0.010025024 1424686735691 1424688581745026978 nexus4_2 145 nexus4 g stand -3.814697E-4 0.0184021 -0.013656616 1424686735890 1424688581945252808 nexus4_2 185 nexus4 g stand -3.814697E-4 -0.031799316 -0.00831604 1424686736094 1424686734097840342 nexus4_1 218 nexus4 g stand -7.324219E-4 -0.013381958 0.01109314 1424686736294 1424688582347932252 nexus4_2 265 nexus4 g stand -0.005722046 0.015197754 0.022659302 1424686736495 1424688582549592408 nexus4_2 305 nexus4 g stand -3.814697E-4 0.0087890625 0.0034332275 1424686736697 1424688582750703248 nexus4_2 345 nexus4 g stand 0.002822876 -0.008300781 -0.015792847 1424686736898 1424688582952241334 nexus4_2 385 nexus4 g stand 6.866455E-4 -0.008300781 0.004501343 1424686737100 1424686735109928643 nexus4_1 418 nexus4 g stand 0.003540039 -0.010177612 -0.026290894 1424686737300 1424688583355164918 nexus4_2 465 nexus4 g stand 0.002822876 0.0045166016 -0.014724731 1424686737505 1424686735512935017 nexus4_1 498 nexus4 g stand 0.0024719238 -0.010177612 -0.017745972 1424686737707 1424686735709254597 nexus4_1 537 nexus4 g stand -0.0028686523 -0.003768921 0.020706177 1424686737908 1424686735915675495 nexus4_1 578 nexus4 g stand -0.0028686523 0.026138306 0.007888794

***b***

In [16]:
staticDF.count()

Out[ 7 ]: 6240991

***c***

In [18]:
staticDF.distinct().count()

Out[ 8 ]: 6240991

There are no duplicates since the .count() and .distinct().count() fetches the same result

***d***

In [21]:
display(staticDF.describe())

summary Arrival_Time Creation_Time Device Index Model User gt x y z count 6240991 6240991 6240991 6240991 6240991 6240991 6240991 6240991 6240991 6240991 mean 1.4247459137335715E12 1.42474684488181734E18 null 174536.91224919248 null null null 4.796918779024442E-4 -0.006015409589636538 -0.010133564891647982 stddev 4.3717662371714726E7 4.372975621925285E13 null 102085.10343046644 null null null 0.452435842800112 0.45913492830935265 0.5232216780091502 min 1424686734958 1424565699954528068 nexus4_1 0 nexus4 a bike -4.128174 -5.791336 -5.2783813 max 1424789460626 1424791305946108620 nexus4_2 396342 nexus4 i walk 9.088684 14.347641 4.597931

***Task 5. GroupedData operations***

***a***

In [24]:
staticDF.groupby(["Model", "Device"]).count().show()

+------+--------+-------+
 Model| Device| count|
+------+--------+-------+
nexus4|nexus4_1|3091811|
nexus4|nexus4_2|3149180|
+------+--------+-------+

In [25]:
staticDF.groupby(["Model", "Device"]).agg(F.count("Device")).show()

+------+--------+-------------+
 Model| Device|count(Device)|
+------+--------+-------------+
nexus4|nexus4_1| 3091811|
nexus4|nexus4_2| 3149180|
+------+--------+-------------+

***b***

In [27]:
#staticDF.groupby("Device", "gt").count().alias("Count_gt_Device").sort("count", ascending = False).show()
staticDF.groupby("Device", "gt").count().sort("count", ascending = False).show()

+--------+----------+------+
 Device| gt| count|
+--------+----------+------+
nexus4_2| walk|534222|
nexus4_1| walk|526180|
nexus4_1| sit|502477|
nexus4_2| sit|482237|
nexus4_2| stand|466886|
nexus4_2| bike|464416|
nexus4_1| stand|443897|
nexus4_1| null|420991|
nexus4_1| stairsup|419926|
nexus4_2| stairsup|416672|
nexus4_2| null|414734|
nexus4_1| bike|399294|
nexus4_1|stairsdown|379046|
nexus4_2|stairsdown|370013|
+--------+----------+------+

In [28]:
staticDF.groupby("Device", "gt").count().sort("count", ascending = False).head(1)

Out[ 30 ]: [Row(Device='nexus4_2', gt='walk', count=534222)]

***c.***
nexus4_2 and walk combination has the highest count with value 534222

***Task 6. Data partitions, repartitioning and shuffle partitions***

In [31]:
staticDF_rdd = staticDF.rdd
staticDF_rdd.getNumPartitions()

Out[ 14 ]: 12

In [32]:
staticDF = staticDF.repartition(100)

In [33]:
staticDF_rdd = staticDF.rdd
staticDF_rdd.getNumPartitions()

Out[ 16 ]: 100

In [34]:
spark.conf.get("spark.sql.shuffle.partitions")

Out[ 17 ]: '200'

In [35]:
spark.conf.set("spark.sql.shuffle.partitions", 20)

In [36]:
spark.conf.get("spark.sql.shuffle.partitions")

Out[ 19 ]: '20'

***Task 7. Create a streaming DataFrame***

***a***

In [39]:
streamingDF = spark\
  .readStream.schema(dataSchema)\
  .option("maxFilesPerTrigger", 1)\
  .format("json")\
  .load("dbfs:/databricks-datasets/definitive-guide/data/activity-data/*.json")

***Task 8. Setting up transformations on Streaming DataFrame***

In [41]:
resultDF = streamingDF.groupby().count()

In [42]:
dbutils.fs.rm("dbfs:/local_disk0/tmp", True)

Out[ 23 ]: True

In [43]:
mystreamingQuery = resultDF\
.writeStream\
.queryName("originaldata")\
.format("memory")\
.outputMode("complete")\
.start()

***Task 10. Viewing information about the streaming query***

***a***

In [46]:
print(mystreamingQuery.status)
print(mystreamingQuery.id)
print(mystreamingQuery.name)
print(mystreamingQuery.isActive)

{'isDataAvailable': True, 'message': 'Getting offsets from FileStreamSource[dbfs:/databricks-datasets/definitive-guide/data/activity-data/*.json]', 'isTriggerActive': True}
76872b9f-b82f-425d-9981-cc8a7adf2cda
originaldata
True

***b***

In [48]:
sqlDF = spark.sql("SELECT * FROM originaldata")

In [49]:
from time import sleep
for x in range(5):
  spark.sql("SELECT * FROM originaldata").show()
  sleep(5)

+-------+
 count|
+-------+
2028305|
+-------+

+-------+
 count|
+-------+
2184331|
+-------+

+-------+
 count|
+-------+
2262344|
+-------+

+-------+
 count|
+-------+
2418372|
+-------+

+-------+
 count|
+-------+
2574400|
+-------+

***c. Stopping the streaming query***

In [51]:
mystreamingQuery.stop()

***Task 11. Calculate aggregate counts of the various combination of values in Device and gt columns using three different approaches***

***groupby***

In [54]:
transformDF1 = streamingDF.groupby(["Device", "gt"]).count()

In [55]:
streamingQuery1 = transformDF1.writeStream\
  .outputMode("complete")\
  .format("memory")\
  .queryName("data1")\
  .start()

In [56]:
from time import sleep
for x in range(5):
  spark.sql("SELECT * FROM data1").show()
  sleep(5)

+--------+----------+-----+
 Device| gt|count|
+--------+----------+-----+
nexus4_1|stairsdown|28543|
nexus4_1| stand|33281|
nexus4_1| sit|37836|
nexus4_2|stairsdown|27649|
nexus4_2| walk|40001|
nexus4_2| null|31159|
nexus4_1| bike|29925|
nexus4_1| stairsup|31431|
nexus4_2| stand|35028|
nexus4_1| null|31529|
nexus4_1| walk|39535|
nexus4_2| bike|34856|
nexus4_2| sit|36019|
nexus4_2| stairsup|31279|
+--------+----------+-----+

+--------+----------+-----+
 Device| gt|count|
+--------+----------+-----+
nexus4_1|stairsdown|33319|
nexus4_1| stand|38726|
nexus4_1| sit|44115|
nexus4_2|stairsdown|32238|
nexus4_2| walk|46608|
nexus4_2| null|36353|
nexus4_1| bike|34857|
nexus4_1| stairsup|36659|
nexus4_2| stand|40968|
nexus4_1| null|36783|
nexus4_1| walk|46184|
nexus4_2| bike|40722|
nexus4_2| sit|42048|
nexus4_2| stairsup|36503|
+--------+----------+-----+

+--------+----------+-----+
 Device| gt|count|
+--------+----------+-----+
nexus4_1|stairsdown|38035|
nexus4_1| stand|44468|
nexus4_1| sit|50411|
nexus4_2|stairsdown|36887|
nexus4_2| walk|53426|
nexus4_2| null|41439|
nexus4_1| bike|39942|
nexus4_1| stairsup|41864|
nexus4_2| stand|46611|
nexus4_1| null|42145|
nexus4_1| walk|52622|
nexus4_2| bike|46435|
nexus4_2| sit|48060|
nexus4_2| stairsup|41750|
+--------+----------+-----+

+--------+----------+-----+
 Device| gt|count|
+--------+----------+-----+
nexus4_1|stairsdown|42835|
nexus4_1| stand|49856|
nexus4_1| sit|56641|
nexus4_2|stairsdown|41451|
nexus4_2| walk|60013|
nexus4_2| null|46671|
nexus4_1| bike|44922|
nexus4_1| stairsup|47065|
nexus4_2| stand|52608|
nexus4_1| null|47362|
nexus4_1| walk|59291|
nexus4_2| bike|52253|
nexus4_2| sit|54137|
nexus4_2| stairsup|47002|
+--------+----------+-----+

+--------+----------+-----+
 Device| gt|count|
+--------+----------+-----+
nexus4_1|stairsdown|47549|
nexus4_1| stand|55451|
nexus4_1| sit|62941|
nexus4_2|stairsdown|46099|
nexus4_2| walk|66900|
nexus4_2| null|51896|
nexus4_1| bike|49989|
nexus4_1| stairsup|52282|
nexus4_2| stand|58398|
nexus4_1| null|52586|
nexus4_1| walk|65660|
nexus4_2| bike|57985|
nexus4_2| sit|60144|
nexus4_2| stairsup|52239|
+--------+----------+-----+

In [57]:
streamingQuery1.stop()

***rollup***

In [59]:
transformDF2 = streamingDF.rollup("Device", "gt").count()

In [60]:
streamingQuery2 = transformDF2.writeStream\
  .outputMode("complete")\
  .format("memory")\
  .queryName("data2")\
  .start()

In [61]:
from time import sleep
for x in range(5):
  spark.sql("SELECT * FROM data2").show()
  sleep(5)

+--------+----------+------+
 Device| gt| count|
+--------+----------+------+
nexus4_1| stand| 49856|
nexus4_2| bike| 52253|
nexus4_2| walk| 60013|
nexus4_2| null|354135|
nexus4_2|stairsdown| 41451|
nexus4_2| sit| 54137|
nexus4_1|stairsdown| 42835|
nexus4_1| sit| 56641|
nexus4_1| bike| 44922|
nexus4_1| null| 47362|
nexus4_2| stairsup| 47002|
nexus4_2| null| 46671|
 null| null|702107|
nexus4_1| null|347972|
nexus4_1| walk| 59291|
nexus4_2| stand| 52608|
nexus4_1| stairsup| 47065|
+--------+----------+------+

+--------+----------+------+
 Device| gt| count|
+--------+----------+------+
nexus4_1| stand| 60882|
nexus4_2| bike| 63880|
nexus4_2| walk| 73340|
nexus4_2| null|432906|
nexus4_2|stairsdown| 50695|
nexus4_2| sit| 66144|
nexus4_1|stairsdown| 52315|
nexus4_1| sit| 69248|
nexus4_1| bike| 54893|
nexus4_1| null| 57829|
nexus4_2| stairsup| 57393|
nexus4_2| null| 57102|
 null| null|858131|
nexus4_1| null|425225|
nexus4_1| walk| 72476|
nexus4_2| stand| 64352|
nexus4_1| stairsup| 57582|
+--------+----------+------+

+--------+----------+------+
 Device| gt| count|
+--------+----------+------+
nexus4_1| stand| 66551|
nexus4_2| bike| 69626|
nexus4_2| walk| 80150|
nexus4_2| null|472227|
nexus4_2|stairsdown| 55355|
nexus4_2| sit| 72126|
nexus4_1|stairsdown| 57016|
nexus4_1| sit| 75573|
nexus4_1| bike| 59946|
nexus4_1| null| 63062|
nexus4_2| stairsup| 62585|
nexus4_2| null| 62317|
 null| null|936143|
nexus4_1| null|463916|
nexus4_1| walk| 78922|
nexus4_2| stand| 70068|
nexus4_1| stairsup| 62846|
+--------+----------+------+

+--------+----------+-------+
 Device| gt| count|
+--------+----------+-------+
nexus4_1| stand| 72118|
nexus4_2| bike| 75510|
nexus4_2| walk| 86661|
nexus4_2| null| 511546|
nexus4_2|stairsdown| 60005|
nexus4_2| sit| 78260|
nexus4_1|stairsdown| 61727|
nexus4_1| sit| 81746|
nexus4_1| bike| 64860|
nexus4_1| null| 68353|
nexus4_2| stairsup| 67749|
nexus4_2| null| 67474|
 null| null|1014155|
nexus4_1| null| 502609|
nexus4_1| walk| 85666|
nexus4_2| stand| 75887|
nexus4_1| stairsup| 68139|
+--------+----------+-------+

+--------+----------+-------+
 Device| gt| count|
+--------+----------+-------+
nexus4_1| stand| 77698|
nexus4_2| bike| 81254|
nexus4_2| walk| 93515|
nexus4_2| null| 550869|
nexus4_2|stairsdown| 64648|
nexus4_2| sit| 84185|
nexus4_1|stairsdown| 66446|
nexus4_1| sit| 88128|
nexus4_1| bike| 69914|
nexus4_1| null| 73656|
nexus4_2| stairsup| 72956|
nexus4_2| null| 72618|
 null| null|1092167|
nexus4_1| null| 541298|
nexus4_1| walk| 92068|
nexus4_2| stand| 81693|
nexus4_1| stairsup| 73388|
+--------+----------+-------+

In [62]:
streamingQuery2.stop()

***cube***

In [64]:
transformDF3 = streamingDF.cube("Device", "gt").count()

In [65]:
streamingQuery3 = transformDF3.writeStream\
  .outputMode("complete")\
  .format("memory")\
  .queryName("data3")\
  .start()

In [66]:
from time import sleep
for x in range(5):
  spark.sql("SELECT * FROM data3").show()
  sleep(5)

+--------+----------+------+
 Device| gt| count|
+--------+----------+------+
nexus4_1| stand| 66551|
nexus4_2| bike| 69626|
nexus4_2| walk| 80150|
 null|stairsdown|112371|
nexus4_2| null|472227|
nexus4_2|stairsdown| 55355|
 null| stand|136619|
 null| bike|129572|
nexus4_2| sit| 72126|
nexus4_1|stairsdown| 57016|
nexus4_1| sit| 75573|
nexus4_1| bike| 59946|
nexus4_1| null| 63062|
nexus4_2| stairsup| 62585|
 null| walk|159072|
nexus4_2| null| 62317|
 null| null|936143|
nexus4_1| null|463916|
 null| null|125379|
nexus4_1| walk| 78922|
+--------+----------+------+
only showing top 20 rows

+--------+----------+-------+
 Device| gt| count|
+--------+----------+-------+
nexus4_1| stand| 77698|
nexus4_2| bike| 81254|
nexus4_2| walk| 93515|
 null|stairsdown| 131094|
nexus4_2| null| 550869|
nexus4_2|stairsdown| 64648|
 null| stand| 159391|
 null| bike| 151168|
nexus4_2| sit| 84185|
nexus4_1|stairsdown| 66446|
nexus4_1| sit| 88128|
nexus4_1| bike| 69914|
nexus4_1| null| 73656|
nexus4_2| stairsup| 72956|
 null| walk| 185583|
nexus4_2| null| 72618|
 null| null|1092167|
nexus4_1| null| 541298|
 null| null| 146274|
nexus4_1| walk| 92068|
+--------+----------+-------+
only showing top 20 rows

+--------+----------+-------+
 Device| gt| count|
+--------+----------+-------+
nexus4_1| stand| 83055|
nexus4_2| bike| 87159|
nexus4_2| walk| 100106|
 null|stairsdown| 140456|
nexus4_2| null| 590415|
nexus4_2|stairsdown| 69195|
 null| stand| 170778|
 null| bike| 161965|
nexus4_2| sit| 90275|
nexus4_1|stairsdown| 71261|
nexus4_1| sit| 94345|
nexus4_1| bike| 74806|
nexus4_1| null| 78903|
nexus4_2| stairsup| 78139|
 null| walk| 198839|
nexus4_2| null| 77818|
 null| null|1170179|
nexus4_1| null| 579764|
 null| null| 156721|
nexus4_1| walk| 98733|
+--------+----------+-------+
only showing top 20 rows

+--------+----------+-------+
 Device| gt| count|
+--------+----------+-------+
nexus4_1| stand| 88791|
nexus4_2| bike| 92898|
nexus4_2| walk| 106915|
 null|stairsdown| 149819|
nexus4_2| null| 629724|
nexus4_2|stairsdown| 73851|
 null| stand| 182165|
 null| bike| 172762|
nexus4_2| sit| 96280|
nexus4_1|stairsdown| 75968|
nexus4_1| sit| 100647|
nexus4_1| bike| 79864|
nexus4_1| null| 84162|
nexus4_2| stairsup| 83400|
 null| walk| 212095|
nexus4_2| null| 83006|
 null| null|1248191|
nexus4_1| null| 618467|
 null| null| 167168|
nexus4_1| walk| 105180|
+--------+----------+-------+
only showing top 20 rows

+--------+----------+-------+
 Device| gt| count|
+--------+----------+-------+
nexus4_1| stand| 94181|
nexus4_2| bike| 98714|
nexus4_2| walk| 113366|
 null|stairsdown| 159182|
nexus4_2| null| 669028|
nexus4_2|stairsdown| 78398|
 null| stand| 193552|
 null| bike| 183559|
nexus4_2| sit| 102357|
nexus4_1|stairsdown| 80784|
nexus4_1| sit| 106878|
nexus4_1| bike| 84845|
nexus4_1| null| 89416|
nexus4_2| stairsup| 88624|
 null| walk| 225351|
nexus4_2| null| 88198|
 null| null|1326203|
nexus4_1| null| 657175|
 null| null| 177614|
nexus4_1| walk| 111985|
+--------+----------+-------+
only showing top 20 rows

In [67]:
streamingQuery3.stop()

***Task 12. Calculate the count of rows where the gt column contains ‘stairs’ only for users a and b on Device=’nexus4_1’***

In [69]:
import pyspark.sql.functions as F
stairsonlyDF = streamingDF.where((F.col("gt").contains("stairs")) & ((F.col("User") == "a") | (F.col("User") == "b")) & (F.col("Device") == "nexus4_1"))
stairsonlyDF = stairsonlyDF.select("User").groupby("User").count()

In [70]:
stairsonlyQuery = stairsonlyDF.writeStream\
  .outputMode("complete")\
  .format("memory")\
  .queryName("stairsonlydata")\
  .start()

In [71]:

for x in range(5):
  spark.sql("SELECT * FROM stairsonlydata").show()
  sleep(5)

+----+-----+
User|count|
+----+-----+
 b| 6717|
 a| 6624|
+----+-----+

+----+-----+
User|count|
+----+-----+
 b| 7867|
 a| 7711|
+----+-----+

+----+-----+
User|count|
+----+-----+
 b| 8952|
 a| 8798|
+----+-----+

+----+-----+
User|count|
+----+-----+
 b|10069|
 a| 9886|
+----+-----+

+----+-----+
User|count|
+----+-----+
 b|11181|
 a|11020|
+----+-----+

In [72]:
stairsonlyQuery.stop()

***Task 13. Compute the average of colummns x, y, x for each combination of values***

In [74]:
gtmodelstatsDF = streamingDF.groupby(["gt", "model"]).agg(F.avg("x").alias("x_avg"), F.avg("y").alias("y_avg"), F.avg("z").alias("z_avg"), F.count("Device").alias("devicecount"))

In [75]:
gtmodelstatsQuery = gtmodelstatsDF.writeStream\
  .format("memory")\
  .outputMode("complete")\
  .queryName("gtmodelstatsdata")\
  .start()

In [76]:
from time import sleep
for x in range(5):
  spark.sql("SELECT * FROM gtmodelstatsdata").where(F.col("x_avg") > 0).show()
  sleep(5)

+----------+------+--------------------+--------------------+--------------------+-----------+
 gt| model| x_avg| y_avg| z_avg|devicecount|
+----------+------+--------------------+--------------------+--------------------+-----------+
stairsdown|nexus4|0.023675364574222757|-0.03823912330772138| 0.12413775718587754| 196623|
 bike|nexus4|0.022988712178907245|-0.01041922896671...|-0.08254010658960201| 226746|
+----------+------+--------------------+--------------------+--------------------+-----------+

+----------+------+--------------------+--------------------+--------------------+-----------+
 gt| model| x_avg| y_avg| z_avg|devicecount|
+----------+------+--------------------+--------------------+--------------------+-----------+
stairsdown|nexus4|0.023415436375415453|-0.03781574129067933| 0.12400820466825611| 205983|
 bike|nexus4|0.022888355830240428|-0.01058633725581...|-0.08265273236428265| 237543|
+----------+------+--------------------+--------------------+--------------------+-----------+

+----------+------+--------------------+--------------------+--------------------+-----------+
 gt| model| x_avg| y_avg| z_avg|devicecount|
+----------+------+--------------------+--------------------+--------------------+-----------+
stairsdown|nexus4| 0.02324887760016299|-0.03718298814411216| 0.12374187743295967| 215342|
 bike|nexus4|0.022932636063705412|-0.01052458422691...|-0.08256997570490414| 248340|
+----------+------+--------------------+--------------------+--------------------+-----------+

+----------+------+--------------------+--------------------+--------------------+-----------+
 gt| model| x_avg| y_avg| z_avg|devicecount|
+----------+------+--------------------+--------------------+--------------------+-----------+
stairsdown|nexus4|0.023002657790226096|-0.03634079193240637| 0.1238389544400663| 234062|
 bike|nexus4|0.022827287245385172|-0.01059932059335...|-0.08266523026675371| 269934|
+----------+------+--------------------+--------------------+--------------------+-----------+

+----------+------+--------------------+--------------------+--------------------+-----------+
 gt| model| x_avg| y_avg| z_avg|devicecount|
+----------+------+--------------------+--------------------+--------------------+-----------+
stairsdown|nexus4|0.023086639781824975|-0.03581268574976...| 0.12370863428123177| 243422|
 bike|nexus4|0.022983098017964537|-0.01066578637587...|-0.08261143843474071| 280731|
+----------+------+--------------------+--------------------+--------------------+-----------+

In [77]:
gtmodelstatsQuery.stop()

***Task 14. Perform a join of a static DataFrame with a streaming DataFrame***

In [79]:
historicalAvg = staticDF.groupby(["gt", "model"]).agg(F.avg("x").alias("x_avg1"), F.avg("y").alias("y_avg1"), F.avg("z").alias("z_avg1"))
historicalAvg.show()

+----------+------+--------------------+--------------------+--------------------+
 gt| model| x_avg1| y_avg1| z_avg1|
+----------+------+--------------------+--------------------+--------------------+
stairsdown|nexus4|0.021613908669165297|-0.03249018824752618| 0.12035922691504061|
 stand|nexus4|-3.11082189691726...|3.218461665975320...|2.141300040636482E-4|
 sit|nexus4|-5.49433244039590...|2.791446281700075E-4|-2.33994461689889...|
 null|nexus4|-0.00847688860109...|-7.30455258739183...|0.003090601491419...|
 bike|nexus4|0.022688759550866834|-0.00877912156368...|-0.08251001663412379|
 walk|nexus4|-0.00390116006094...|0.001052508689953...|-6.95435553042996...|
 stairsup|nexus4|-0.02479965287771...|-0.00800392344379...|-0.10034088415060397|
+----------+------+--------------------+--------------------+--------------------+

In [80]:
streamingAvg = streamingDF.groupby(["gt", "model"]).agg(F.avg("x").alias("x_avg2"), F.avg("y").alias("y_avg2"), F.avg("z").alias("z_avg2"))

In [81]:
jointransformDF = historicalAvg.join(streamingAvg, on = ["gt", "model"])

In [82]:
joinQuery = jointransformDF.writeStream\
  .format("memory")\
  .outputMode("complete")\
  .queryName("jointransformdata")\
  .start()

In [83]:
for x in range(5):
  spark.sql("SELECT * FROM jointransformdata").show()
  sleep(5)

+----------+------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+
 gt| model| x_avg1| y_avg1| z_avg1| x_avg2| y_avg2| z_avg2|
+----------+------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+
stairsdown|nexus4|0.021613908669165297|-0.03249018824752618| 0.12035922691504061|0.028103791071500104|-0.03570080351911373| 0.12203047970606548|
 stand|nexus4|-3.11082189691726...|3.218461665975320...|2.141300040636482E-4|-3.00989804137386...|4.133303473120163...|-2.86960196767402...|
 sit|nexus4|-5.49433244039590...|2.791446281700075E-4|-2.33994461689889...|-4.92132825379803...|3.757376157445784...|-4.42863346250729...|
 null|nexus4|-0.00847688860109...|-7.30455258739183...|0.003090601491419...|-0.00302501221506...|-0.00410754501410...|0.005961452067049477|
 bike|nexus4|0.022688759550866834|-0.00877912156368...|-0.08251001663412379|0.023512544470933663|-0.01304747996973...|-0.08360475809007027|
 walk|nexus4|-0.00390116006094...|0.001052508689953...|-6.95435553042996...|0.001970352086338262|7.489666845957323E-5|-0.00149828380428...|
 stairsup|nexus4|-0.02479965287771...|-0.00800392344379...|-0.10034088415060397|-0.02623301318863378| -0.0138593176529181| -0.093950097280195|
+----------+------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+

+----------+------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+
 gt| model| x_avg1| y_avg1| z_avg1| x_avg2| y_avg2| z_avg2|
+----------+------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+
stairsdown|nexus4|0.021613908669165297|-0.03249018824752618| 0.12035922691504061|0.028103791071500104|-0.03570080351911373| 0.12203047970606548|
 stand|nexus4|-3.11082189691726...|3.218461665975320...|2.141300040636482E-4|-3.00989804137386...|4.133303473120163...|-2.86960196767402...|
 sit|nexus4|-5.49433244039590...|2.791446281700075E-4|-2.33994461689889...|-4.92132825379803...|3.757376157445784...|-4.42863346250729...|
 null|nexus4|-0.00847688860109...|-7.30455258739183...|0.003090601491419...|-0.00302501221506...|-0.00410754501410...|0.005961452067049477|
 bike|nexus4|0.022688759550866834|-0.00877912156368...|-0.08251001663412379|0.023512544470933663|-0.01304747996973...|-0.08360475809007027|
 walk|nexus4|-0.00390116006094...|0.001052508689953...|-6.95435553042996...|0.001970352086338262|7.489666845957323E-5|-0.00149828380428...|
 stairsup|nexus4|-0.02479965287771...|-0.00800392344379...|-0.10034088415060397|-0.02623301318863378| -0.0138593176529181| -0.093950097280195|
+----------+------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+

+----------+------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+
 gt| model| x_avg1| y_avg1| z_avg1| x_avg2| y_avg2| z_avg2|
+----------+------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+
stairsdown|nexus4|0.021613908669165297|-0.03249018824752618| 0.12035922691504061|0.025383337512333622| -0.0362580763332372| 0.12712694345857248|
 stand|nexus4|-3.11082189691726...|3.218461665975320...|2.141300040636482E-4|-3.19865608911239...|4.070066943871052E-4|1.027885098598942...|
 sit|nexus4|-5.49433244039590...|2.791446281700075E-4|-2.33994461689889...|-4.84874225480320...|3.237740430277438...|-4.65213713798297E-5|
 null|nexus4|-0.00847688860109...|-7.30455258739183...|0.003090601491419...|-0.00601179783148...|-7.22352747272202...|0.003847526633724...|
 bike|nexus4|0.022688759550866834|-0.00877912156368...|-0.08251001663412379|0.020961064903431637|-0.009152226

In [84]:
joinQuery.stop()

The values of x_avg2, y_avg2, z_avg2 changes over time.